src/content/blog/search-notes.md

```diff
@@ -0,0 +1,71 @@
+
+### pg_trgrm
+
+The solution that we've soft-launched as the v2 search algorithm is based on [`pg_trgrm`](https://www.postgresql.org/docs/current/pgtrgm.html), which implements [trigram search](https://en.wikipedia.org/wiki/Trigram_search) in Postgres. Code search _does_ seem to succeed with trigrams: [Russ Cox's famous (to me?) piece from 2012 tells the story of how Google Code Search](https://swtch.com/~rsc/regexp/regexp4.html) used trigram indexes and a special regex implementation to succeed, technically. GitHub's [new search system](https://github.blog/2023-02-06-the-technology-behind-githubs-new-code-search/) uses trigram search too, in addition to a lot of technology that I'm jealous of. [Sourcegraph](https://github.com/sourcegraph/zoekt) has a trigram-based search tool that they've inherited from Google, too.
+
+Our work with the Postgres `pg_trgrm` approach has been heavily informed by [Stephen Gutekanst’s blog post series about indexing repositories locally](https://devlog.hexops.com/2021/postgres-regex-search-over-10000-github-repositories/) in Postgres. We've created a [GIN](https://www.postgresql.org/docs/current/gin.html) index with `gin_trgm_ops` on a column containing search text.
+
+The conclusion so far is that this is a great solution for regex search, but we're not doing regex search: most searches are more freeform. We're using [word_similarity](https://www.postgresql.org/docs/current/pgtrgm.html#PGTRGM-FUNCS-OPS) for search ranking, and it has been very hard to coax the algorithm into giving us anything like a reasonable ranking.
+
+### The universe of options
+
+| Option                                        | Architecture       | Language | Stars |
+| --------------------------------------------- | ------------------ | -------- | ----- |
+| [Meilisearch](https://www.meilisearch.com/)   | Standalone         | Rust     | 41k   |
+| [Typesense](https://typesense.org/)           | Standalone         | C++      | 17k   |
+| [Zoekt](https://github.com/sourcegraph/zoekt) | Standalone         | Go       | 406   |
+| [ParadeDB](https://www.paradedb.com/)         | Postgres extension | Rust     | 3.2k  |
+
+There are code-specific tools that exist, but most of them are closed-source: GitHub's search is excellent, but is obviously the work of a dedicated team with a real time budget.
+
+- Sourcegraph's maintained fork of [Zoekt](https://github.com/google/zoekt) is pretty cool, but is pretty fearfully niche and would be a big, new infrastructure commitment.
+- [Elasticsearch](https://github.com/elastic/elasticsearch) might be the eventual, unavoidable solution to this problem. It doesn't have code-specific handling, but can be tuned in nearly infinite ways. We're not excited to start learning about Java memory tuning and to introduce the first persistent disk storage to our application, as well as an additional source of truth for our data.
```

Comment:
Double use of "tune" but meaning different things

Reply:
The java stuff feels a bit like a straw man, wouldn't we use elastic search as a service?